modal/_utils/async_utils.py

```diff
@@ -460,3 +460,48 @@ def run_generator_sync(
         except BaseException as err:
             exc = err
     loop.close()
+
+
+class DynamicSemaphore:
+    def __init__(self, capacity: int):
+        self._capacity = capacity
+        self._new_capacity = capacity
+        self._semaphore = asyncio.Semaphore(capacity)
+        self._owed_releases = 0
+        self._lock = asyncio.Lock()
+
+    def get_capacity(self) -> int:
+        return self._capacity
+
+    async def acquire(self) -> None:
+        async with self._lock:
+            await self._update_capacity()
+        await self._semaphore.acquire()
+
+    async def acquire_all(self) -> None:
+        async with self._lock:
+            await self._update_capacity()
+            for _ in range(self._capacity):
+                await self._semaphore.acquire()
+
+    def release(self):
+        if self._owed_releases > 0:
+            self._owed_releases -= 1
+        else:
+            self._semaphore.release()
+
+    def set_capacity(self, new_capacity: int):
+        self._new_capacity = new_capacity
+
+    async def _update_capacity(self):
+        if self._capacity < self._new_capacity:
+            for _ in range(self._new_capacity - self._capacity):
+                self.release()
+        elif self._capacity > self._new_capacity:
+            for _ in range(self._capacity - self._new_capacity):
+                if self._semaphore.locked():
+                    self._owed_releases += 1
+                else:
+                    await self._semaphore.acquire()
```

Comment:
there's actually a race here with `self._semaphore.acquire()` in `acquire` (line 479), might have to change this to `self._semaphore._value -= 1` to make the entire function sync